### PR TITLE
New graph visitor

### LINKED ISSUE #216: Tree layout doesn't support nested parallel stages
### Jenkins and plugins versions report

<details>
  <summary>Environment</summary>

  
  ```text
Jenkins: 2.361.4
OS: Windows 11 - 10.0
Java: 11.0.17 - Eclipse Adoptium (OpenJDK 64-Bit Server VM)
---
ace-editor:1.1
apache-httpcomponents-client-4-api:4.5.13-138.v4e7d9a_7b_a_e61
bouncycastle-api:2.26
branch-api:2.1051.v9985666b_f6cc
caffeine-api:2.9.3-65.v6a_47d0f4d1fe
cloudbees-folder:6.758.vfd75d09eea_a_1
command-launcher:1.2
credentials:1189.vf61b_a_5e2f62e
credentials-binding:523.vd859a_4b_122e6
dark-theme:262.v0202a_4c8fb_6a
display-url-api:2.3.6
durable-task:500.v8927d9fd99d8
git:4.14.1
git-client:3.13.0
github:1.34.3
github-api:1.303-400.v35c2d8258028
github-branch-source:1696.v3a_7603564d04
instance-identity:116.vf8f487400980
ionicons-api:31.v4757b_6987003
jackson2-api:2.13.4.20221013-295.v8e29ea_354141
jakarta-activation-api:2.0.1-2
jakarta-mail-api:2.0.1-2
javax-activation-api:1.2.0-5
javax-mail-api:1.6.2-5
jaxb:2.3.7-1
jdk-tool:1.0
jjwt-api:0.11.5-77.v646c772fddb_0
jsch:0.1.55.61.va_e9ee26616e7
mailer:438.v02c7f0a_12fa_4
metrics:4.2.10-405.v60a_9cc74e923
okhttp-api:4.9.3-108.v0feda04578cf
pipeline-graph-analysis:195.v5812d95a_a_2f9
pipeline-graph-view:999999-SNAPSHOT (private-1abdd280-stuartr)
pipeline-groovy-lib:621.vb_44ce045b_582
pipeline-input-step:466.v6d0a_5df34f81
pipeline-model-api:2.2118.v31fd5b_9944b_5
pipeline-model-definition:2.2118.v31fd5b_9944b_5
pipeline-model-extensions:2.2118.v31fd5b_9944b_5
pipeline-stage-step:296.v5f6908f017a_5
pipeline-stage-tags-metadata:2.2118.v31fd5b_9944b_5
plain-credentials:139.ved2b_9cf7587b
scm-api:621.vda_a_b_055e58f7
script-security:1218.v39ca_7f7ed0a_c
snakeyaml-api:1.33-90.v80dcb_3814d35
ssh-credentials:305.v8f4381501156
sshd:3.236.ved5e1b_cb_50b_2
structs:324.va_f5d6774f3a_d
theme-manager:1.5
token-macro:321.vd7cc1f2a_52c8
trilead-api:2.84.v72119de229b_7
variant:59.vf075fe829ccb
workflow-api:1200.v8005c684b_a_c6
workflow-basic-steps:994.vd57e3ca_46d24
workflow-cps:3536.vb_8a_6628079d5
workflow-durable-task-step:1217.v38306d8fa_b_5c
workflow-job:1254.v3f64639b_11dd
workflow-multibranch:716.vc692a_e52371b_
workflow-scm-step:400.v6b_89a_1317c9a_
workflow-step-api:639.v6eca_cd8c04a_a_
workflow-support:839.v35e2736cfd5c
  ```

</details>

### What Operating System are you using (both controller, and any agents involved in the problem)?

```
OS: Windows 11 - 10.0
```

### Reproduction steps

This issue can be reproduced with the example pipeline below:
```groovy
node {
    parallel([
        "A": {
            stage("Build") {
                echo("Build A")
            }
            stage("Test") {
                parallel([
                    "A1" : {
                       echo("Test A1")
                    },
                    "A2" : {
                       echo("Test A2")
                    }
                ])
            }
        },
        "B": {
            stage("Build") {
                echo("Build B")
            }
            stage("Test") {
                parallel([
                    "B1" : {
                       echo("Test B1")
                    },
                    "B2" : {
                       echo("Test B2")
                    }
                ])
            }
        }
    ])
}
```

### Expected Results

The parallel branches for `A1`, `A2` and `B1`, `B2` should be nested under the `Test` stage of the parent `A` and `B` branches respectively. The pipeline steps view provided by the [Pipeline: Supporting APIs Plugin](https://github.com/jenkinsci/workflow-support-plugin) correctly handles this:
image

### Actual Results

The nested parallel branches are incorrectly shown as siblings of their parent parallel branch:
image

### Anything else?

May be similar / duplicate of https://github.com/jenkinsci/pipeline-graph-view-plugin/issues/51, but I believe that was created before the move to the tree layout and doesn't include the pipeline console view issue.

### LINKED ISSUE #222: Propagate status of nested stages to parent stage in console view
### What feature do you want to see added?

When working with parallel or otherwise with sub-stages, e.g.:
```
pipeline 
{
    agent any
    
    stages 
    {
        stage('ParentStage') 
        {
            stages
            {
                stage('SubStageA')    
                {
                    steps
                    {
                        echo 'SubStageA'
                        bat 'exit 1'
                    }
                }
                
                stage('SubStageB')    
                {
                    steps
                    {
                        echo 'SubStageB'
                    }
                }
            }
        }
    }
}
```
the Pipeline Console view looks like this when a sub-stage fails:
grafik

We expected that the ParentStage is also marked as failed in this case and think that it would be a nice improvement to do so. Or is there a valid reason for the current behavior?

### Upstream changes

_No response_

### PR DESCRIPTION
**Pull request by timbrown5**
Update PipelineStepVisitor to (almost) replace PipelineNodeGraphVisitor.
I think the implementation is easier to understand - nodes are parsed when we hit a start block, not when we child the child start block - I found getting parent often returned null (which might be the cause of some of the issues).
@mikedld is this sort of what you were thinking of in: https://github.com/jenkinsci/pipeline-graph-view-plugin/pull/231#issuecomment-1445064695 ?

Known Issues:
- [ ] PipelineStepVisitor gets called twice in PipelineStepsApi - once for Stages and once for steps. Could we store a Visitor for a session (or would that make it harder to update)?
- [x] New implementation adds additional nodes to the graph (Parallel Branches). This breaks some tests (but doesn't seem to affect 
- [x] the new PipelineStepVisitor incorrectly sets the stage if parallel blocks to not_built. I think this is the cause of some of the unit test failures. In truth setting the correct state of the NodeNodeWrapper is surprisingly tricky.
- [x] Not sure if this is fixed completely, but changing the order of status checks in 'handleChunkDone' has resolved some of the issues I was seeing.

Seems to fix #216  (added test for this):
Screenshot 2023-03-01 at 19 30 28

But not #51 (could be some updates require on the frontend?):
Screenshot 2023-03-01 at 19 30 53

Pipeline:
```
node {
    parallel([
        "A": {
            stage("Build") {
                echo("Build A")
            }
            stage("Test") {
                parallel([
                    "A1" : {
                       echo("Test A1")
                    },
                    "A2" : {
                       echo("Test A2")
                    }
                ])
            }
        },
        "B": {
            stage("Build") {
                echo("Build B")
            }
            parallel([
                "B1" : {
                   echo("Test B1")
                },
                "B2" : {
                   echo("Test B2")
                }
            ])
        }
    ])
}
```

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by stuartrowe** [comment]
@timbrown5 thanks! This seems to fix https://github.com/jenkinsci/pipeline-graph-view-plugin/issues/216 as well.
image

**Comment by timbrown5** [comment]
@stuartrowe thanks for that 👍. I guess I should add a test for that to make sure it doesn't get broken in the future.
My belief if that this gets a accurate representation of the node graph. The issue is that the other classes rely on the behaviour of the PipelineNodeGraphVisitor, which seems to have grown organically and so creates a slightly esoteric representation. Parallel branches are a good example of this:
- In the graph there is a specific node for this, but the PipelineGraphNodeViaitor ignored this and either:
Added blocks to the parent stage (if it exists).
- (If the parallel block isn't wrapped in a stage) add a synthetic stage to wrap the parallel block in.
This seems needless as you can just return the parallel branches wrapped in the parallel block.
Unfortunately, the new representation means updating some of the other classes.

**Comment by timbrown5** [comment]
With the latest change, the graph looks better with a Nested parallel blocks (the stages aren't missing), but still not correct (they aren't branched, but are children):
Screenshot 2023-03-03 at 09 59 06

**Comment by timbrown5** [comment]
FYI, this doesn't fix the Pipeline Graph for nested Pipelines:
Screenshot 2023-03-06 at 18 31 52
Screenshot 2023-03-06 at 18 31 56

File: src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineNodeGraphAdapter.java [hunk 1]
```diff
@@ -129,49 +183,65 @@ public List<FlowNodeWrapper> getPipelineNodes() {
...
 
-  public Map<String, List<FlowNodeWrapper>> getAllSteps() {
-    Map<String, List<FlowNodeWrapper>> stepsMap = treeVisitor.getAllSteps();
-    Map<String, String> remappedStages = getNodesToRemap();
-    for (String originalParentId : remappedStages.keySet()) {
-      if (stepsMap.containsKey(originalParentId)) {
-        String remappedParentId = remappedStages.get(originalParentId);
+
+  private void remapStepParentage() {
+    // We only want to do this once, so return early if we jhave an existing value.
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineNodeGraphAdapter.java:199]:
```suggestion
    // We only want to do this once, so return early if we have an existing value.
```

**Comment by timbrown5** [comment]
I'm leaving this as a draft as I don't know what additional checks would be required to land.

I'm, also a little torn with changing (breaking?) the Node Graph - I could revert it to still use `PipelineNodeGraphVisitor` (and move the changes to `PipelineGraphApi` to a new class - probably `PipelineTreeApi`.
Then the console via and graph via can call different Visitors - and the GraphView moved at a later point if/when the frontend code is updated.

**Comment by timja** [comment]
no real opinion at this point, whatever you think is best, test coverage gives confidence in these areas.

**Comment by timbrown5** [comment]
I think the rest checks the nodes are structured correctly but not that they are the same type.
The types are probably different which might be the cause of the different behaviour.

**Comment by timbrown5** [comment]
I think fixing the Graph front-end for nested parallel stages can be tracked in #51.
I have added an example Pipeline and how I think it should be presented (although I could be wrong)
https://github.com/jenkinsci/pipeline-graph-view-plugin/issues/51#issuecomment-1460169695

**Pull request merged**